examples/test.py

```diff
@@ -0,0 +1,117 @@
+import hydra
+
+from omegaconf import DictConfig
+from omegaconf import OmegaConf
+from typing import List
+
+from enum import Enum
+from dataclasses import dataclass, field
+
+#@dataclass
+class NeMoConfig(dict):
+    """ Abstract NeMo Configuration class"""
+    pass
+    def serialize(self):
+        """ Maybe we will have it in here - or maybe not ;) """
+        pass
+
+    def deserialize(self):
+        """ Maybe we will have it in here - or maybe not ;) """
+        pass
+
+class Height(Enum):
+    """ Enum - just to test how enums can be nested """
+    SHORT = 0
+    TALL = 1
+
+@dataclass
+class ModuleConfig(NeMoConfig):
+    """ This is a module config  - just to test the inheritance """
+    name: str="my_name"
+    val: int=1
+    height: Height = Height.SHORT # Enum!
+
+@dataclass
+class OptimConfig(NeMoConfig):
+    """ This is an optimizer config """
+    name: str="sgd"
+    lr: float=0.99
+
+
+@dataclass
+class JasperBlockConfig(NeMoConfig):
+    """ Default config of a single Jasper block """
+    dilation: int=1
+    dropout: float=0.0
+    filters: int=256
+    kernel: int=33
+    repeat: int=5
+    residual: bool=True
+    separable: bool=True
+    stride: int=1
+
+
+
+@dataclass
+class JasperEncoderConfig(ModuleConfig):
+    """ Config of Jasper encoder """
+    activation: str="relu"
+    conv_mask: bool=True
+    feat_in: int=64
+
+    def __init__(self, **kwargs):
+        """ Init - adds additional parameters, currently - adds 10 JasperBlocks """
+        super().__init__(self, kwargs)
+        # Add blocks.
+        self.blocks = []
+        # Add first block with some custom params.
+        self.blocks.append(JasperBlockConfig(repeat=1, residual=False, stride=2))
+
+        # Add remaining 9 using default params.
+        for i in range(9):
+            self.blocks.append(JasperBlockConfig())
+
+
+@dataclass
+class ModelConfig(NeMoConfig):
+    """ This is a module config  - just to test the inheritance """
+    name: str="my_name"
+
+@dataclass
+class JasperConfig(ModelConfig):
+    # Encoder.
+    encoder: JasperEncoderConfig = field(default_factory=JasperEncoderConfig)
+    # Optimizer.
+    opt: OptimConfig = field(default_factory=OptimConfig)
+
+# Jasper "Prototype" ;)
+def MyJasper(conf: JasperConfig=JasperConfig()):
+    print("="*80 + " Config objects " + "="*80)
+    print("Model name: ", conf.name)
+    print("Model optimizer params: ", conf.opt)
+    print("Model encoder activation: ", conf.encoder.activation)
+    #print("Model encoder params: ", conf.encoder)
+    print("Model encoder blocks: ", conf.encoder.blocks)
+
+
+
+@hydra.main(config_path="test.yaml")
```

Comment:
Hardcoded config file? Can we pass a config file from command line ?

Reply:
No, this is how it's done with Hydra.

Reply:
And config groups require a hierarchical structure of config files ? As in it can't be done with a single config file ?

Reply:
For Hydra you pass in one config file. With that config file "test.yaml" in this example, you can pass in arbitrary many config files (taking the place of passing configs through argparse). But config groups is recommended rather thatn passing in arbitrary config files.

Reply:
Having a hard coded file name that the user must always create exactly with same names reduces flexibility - they can't easily swap out the master config. But I guess that's just another limitation of hydra which we have to consider.

Reply:
hey @titu1994 please note it is work in progress and there are two, not really compatible (yet) solutions.

1) I am learning how to use "pure hydra" (thus @hydra.main(config_path="test.yaml")
2) I am figuring out the best way to write configuration dataclasses

Next step is to merge those two - and getting rid of "hardcoded file names" is honestly my goal.

Please note at that point we won't be "pure hydra" anymore. What I really want to take from hydra is the possibility to override config params from command line

Reply:
@tkornuta-nvidia Yup, that makes complete sense. Hopefully a middle ground can be reached which offers the best of both worlds (under reasonable assumptions and conditions).

Reply:
On that note, I think "Structured Config Schema" https://hydra.cc/docs/next/tutorials/structured_config/schema is probably our best bet for merging the two. And it fits in really nicely with NeMo type checking.  

So for a simple app, we have one config file with everything. As soon as we want to use multiple models, optimizers, datasets, we are using config groups (multiple config files). Finally, all config files and command line overrides will be checked against all of our Nemo Structured Configs.

Reply:
I've been trying to bing the config objects in OmegaConf:
https://omegaconf.readthedocs.io/en/latest/structured_config.html
as this "configuration object with runtime type safety" seems to be well aligned with NeMo Typing.

Has anyone experience with this package?

Reply:
I have a concern with data classes. It's dependent on Python 3.7, whose adoption is still quite low. We need a similar solution to data classes which supports 3.6 at the least, possibly even 3.5.

Reply:
hmmmm, ok, so this is a functional requirement that I first hear of.

I like those dataclasses. And OmegaConf. And as far as I understand hydra depends on both of them...

If I understand it correctly, dataclasses some time ago got backward compatibility for Python 3.6
https://pypi.org/project/dataclasses/
but please correct me if I get it wrong...

@okuchaiev what do you think about this?

Reply:
ps. NeMo readme page clearly says: Python 3.6 or 3.7

So I guess 3.5 is out of our consideration. Phew! @titu1994 :)

Reply:
3.5 is out. 3.6 is not and that doesn't support dataclass.

Reply:
Oh to be clearer, there's now a pip installable dataclass, which is language version independent, backported for 3.6. That is a reasonable dependency then.

Reply:
Phew! Thanks, @titu1994 , my blood pressure goes back to normal... ;)

Reply:
I feel like we can require Python >=3.6